---
title: "Introduction to BayesMultiomics Package"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{BayesMultiomics}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, echo=FALSE, results=FALSE, message=FALSE}
library(BayesMultiomics)
```

## Outline

- Purpose of the package
- Brief explanation of the algorithms and workflow
- description of example data used
- DAVID functional classification tool
- 1st stage modeling with `FSDIM()`
- 2nd stage modeling with `SSDIM()`
- summary of the model results and plots

## 1. Purpose

With the increasing availability of biomedical data from multiple platforms of the same patients in clinical research, such as epigenomics, gene expression, and clinical features, there is a growing need for statistical methods that can jointly analyze data from different platforms to provide complementary information for clinical studies. EMMultiOmics package utilizes a two-stage hierarchical Bayesian model that integrates high-dimensional biomedical data from diverse platforms to select biomarkers associated with clinical outcomes of interest.

In the first stage, we use Expectation Maximization based approach to learn the regulating mechanism between epigenomics (e.g., gene methylation) and gene expression while considering functional gene annotations. In the second stage, we group genes based on the regulating mechanism learned in the first stage. Then, we apply a group-wise penalty to select genes significantly associated with clinical outcomes while incorporating clinical features. 

## 2. Algorithms

(need details..)

![Baisc workflow](../inst/img/workflow.png)

![Model diagram](../inst/img/model_diagram.png)

## 3. example data

The package exports an example data - a Glioblastoma (GBM) dataset (from...). The example gene symbols used are exported in an object `eg_gene_symbols.` To obtain gene grouping, we need to get functional classification of the genes using DAVID tool.

(Also give a brief overview of the dataset if possible. e.g. clinical info, methylation, Gene expression, Y, ...)

## 4. David functional classifcation tool

`vignette("gene_grouping")`

Once the classification file is downloaded as txt, use `DAVIDGeneGrouping()` function to obtain the gene grouping.

```{r, eval=FALSE}
fp <- system.file("eg_fx_classification.txt", package="BayesMultiomics")
gene_grouping <- DAVIDGeneGrouping(eg_gene_symbols, fp)
gene_grouping
```

## 5. First Stage Data Integration Model (FSDIM)

In the first stage, we will combine information from methylation, gene expression, and gene
function annotation data to group genes based on how their expressions are regulated by DNA
methylation.

FSDIM aims to combine information from methylation `M`, gene expression `G`, and gene function annotation data (functional classification) to group genes based on how their expressions are regulated by DNA methylation. Essentially, Essentially, we divide the given K genes into three groups: (1) genes that are modulated by only DNA methylation (M type effect); (2) genes that are modulated by aspects other than DNA methylation (MÂ¯ type effect); (3) genes that are modulated jointly by DNA methylation and aspects other than DNA methylation.Genes will be assigned group membership depending on the coefficient of determination, R2 of FSDIM.

The functional classification used in FDSIM is obtained from Database for Annotation, Visualization, and Integrated Discovery ([DAVID](https://david.ncifcrf.gov)). 

(example code on how to use `FSDIM`)
```{r, eval=FALSE}
# example data
M <- GBM_data2$M
G <- GBM_data2$G

FSDIM_result <- FSDIM(M, G, gene_grouping, I=10, thresh=0.001)
```

Classify genes into 3 groups using `Zmat_builder`


## 6. Second Stage Data Integration Model (SSDIM)

SSDIM is a linear model that can be used to detect biomarkers associated with the clinical outcome `Y` of interest while considering the knowledge learned from FSDIM. It regresses clinical outcome of interest on relevant clinical features `C` and gene expression `G`, regularized with a group penalty reflecting methylation regulating mechanism and gene functional similarity learned from FSDIM. In this way, the SSDIM further integrates clinical outcomes to develop a unified way of combining and curating the set of predictor variables with the clinical responses.

(example code on how to use `NEG_em`, may change as necessary)
```{r, eval=FALSE}
G <- GBM_data2$G
Y <- GBM_data2$Y
C <- GBM_data2$C
R2 <- FSDIM_result$R2
Zmatrix <- Zmat_builder(R2, G, 0.2, 0.8)
mpmath <- setup_mpmath()
a0 <- 0.1
gstr <- "scale"

SSDIM_result <- SSDIM(Y, G, C, a0, gstr, Zmatrix,
                      I=10, thresh=0.001, .mpmath=mpmath)
```

## 7. Data Analysis with `Bayesmultiomics`

`BayesMultiomics` function puts together the first and second stage models and performs the entire algorithm. The output summarizes 1) selected genes from the first stages modeling, 2) gene functional classificaiton groups and methylation groups; 3) coefficients of the clinical features and selected genes. `BayesMultiomics.CV` is an extended version `BayesMultiomics` where the performance of models are evaluated using K-fold cross validation. `BayesMultiomicsPriorSensitivity` function performs the same tasks over a range of values of priors `a` and `g`. 

Example with `BayesMultiomics`
```{r eval=FALSE}
M <- GBM_data2$M
G <- GBM_data2$G
fp <- system.file("eg_fx_classification.txt", package="BayesMultiomics")
gene_grouping <- DAVIDGeneGrouping(eg_gene_symbols, fp)
Y <- GBM_data2$Y
C <- GBM_data2$C
a0 <- 0.1
gstr <- "scale"
mpmath <- setup_mpmath()

# run
multiomics <- BayesMultiomics(
  M=M,
  G=G,
  grouping=gene_grouping,
  Y=Y,
  C=C,
  a0=a0,
  gstr=gstr,
  EMVS_I = 10,
  NEG_I= 10,
  EMVS_thresh = 0.0001,
  NEG_thresh = 0.0001,
  transform_M = "L",
  lower_R2 = 0.2,
  upper_R2 = 0.8,
  .mpmath = mpmath
)
```

```{r echo=FALSE}
# pre-run results
# eg_multiomics <- multiomics
multiomics <- BayesMultiomics::eg_multiomics
```

```{r fig.width=8, fig.height=8}
# summaries
print(multiomics)
```


Example on `BayesMultiomics.CV`
```{r eval=FALSE}
M <- GBM_data2$M
G <- GBM_data2$G
fp <- system.file("eg_fx_classification.txt", package="BayesMultiomics") # replace with your own file path
gene_grouping <- DAVIDGeneGrouping(eg_gene_symbols, fp)
Y <- GBM_data2$Y
C <- GBM_data2$C
a0 <- 0.1
gstr <- "scale"
Delta <- GBM_data2$Delta

# run
multiomics_cv <- BayesMultiomics.CV(
  M=M,
  G=G,
  grouping=gene_grouping,
  Y=Y,
  C=C,
  a0=a0,
  gstr=gstr,
  Delta=Delta,
  n_fold=10,
  random_seed = 123,
  EMVS_I = 10,
  NEG_I=20,
  EMVS_thresh = 0.0001,
  NEG_thresh = 0.0001
  )
```
```{r echo=FALSE}
# pre-run results
# eg_multiomics_cv <- multiomics_cv
multiomics_cv <- BayesMultiomics::eg_multiomics_cv
```

```{r fig.width=8, fig.height=8}
print(multiomics_cv)
plot(multiomics_cv)
```

Example on `BayesMultiomicsPriorSensitivity`
```{r eval=FALSE}
# params
M <- GBM_data2$M
G <- GBM_data2$G
fp <- system.file("eg_fx_classification.txt", package="BayesMultiomics") # replace with your own file path
gene_grouping <- DAVIDGeneGrouping(eg_gene_symbols, fp)
Y <- GBM_data2$Y
C <- GBM_data2$C
Delta <- GBM_data2$Delta

a0 <- c(0.1, 1, 10, 50)
g <- list("scale", 1)
# run
multiomics_sensitivity <- BayesMultiomicsPriorSensitivity(
  M=M,
  G=G,
  grouping=gene_grouping,
  Y=Y,
  C=C,
  a0_vec=a0,
  gstr_vec=g,
  Delta=Delta,
  n_fold=10,
  random_seed = 123,
  EMVS_I = 10,
  NEG_I=20,
  EMVS_thresh = 0.0001,
  NEG_thresh = 0.0001
  )
```

```{r echo=FALSE}
# pre-run results
# eg_multiomics_sensitivity <- multiomics_sensitivity
# save all of them
# save(eg_multiomics, eg_multiomics_cv, eg_multiomics_sensitivity,
#      file="../data/eg_multiomics.rda")
multiomics_sensitivity <- BayesMultiomics::eg_multiomics_sensitivity
```

```{r fig.width=8, fig.height=8}
# summary
summary(multiomics_sensitivity)
plot(multiomics_sensitivity)
```

(walkthrough on how to use `multiOmics` and `multiOmics_sensitivity` to create summary output/tables/plots...)

